### ISSUE TITLE
Gaps aggregator does not fill the whole requested time range

### ISSUE THREAD

**Issue by francisdb** [kairosdb/kairosdb#339]
My query looks like this

from: `2016-10-01 12:00:00.000 am`
to : `2016-10-26 12:00:00.000 am`
sampled gaps/avg by day

I only get 12 results starting at`2016-10-17` (since we only logged this data from the 17th on)

I would expect the gaps aggregator to also fill the gap from the start of the requested range until the first datapoint.

If this is by design is there an other aggregator (like fill) that would give me the expected result?

the query for reference

```
{
  "metrics": [
    {
      "tags": {
        "resource": [
          "myserver"
        ]
      },
      "name": "metric.cpuUsage",
      "aggregators": [
        {
          "name": "avg",
          "align_sampling": true,
          "sampling": {
            "value": "1",
            "unit": "days"
          },
          "align_start_time": true
        },
        {
          "name": "gaps",
          "align_sampling": true,
          "sampling": {
            "value": "1",
            "unit": "days"
          },
          "align_start_time": true
        }
      ]
    }
  ],
  "cache_time": 0,
  "start_absolute": 1475272800000,
  "end_absolute": 1477432800000
}
```

**Comment by brianhks** [comment]
I'm not sure gaps aggregator is what you want.  The aggregator creates actual gaps in the graph like this
![gaps](https://cloud.githubusercontent.com/assets/2686690/21599741/c54fd1e0-d12e-11e6-845b-3d821d6b6d33.png)
It sounds like you want to insert zeros where you have no data.  We could either enhance the gaps agg to take a param that it fills with instead of null or create a new fill aggregator.

**Comment by francisdb** [comment]
The docs mention:

> Marks gaps in data according to sampling rate with a null data point

It seems that this is only done for gaps within existing datapoints. We would need the `gaps?` at the start / end of the requested range to also be filled with `null` datapoints.

**Comment by lcoulet** [comment]
Hi Francis, this is by design. 
The purpose is to make sure that when there's a data dropout, the plotting libraries  should show a drop. 
At the beginning and the end of the query there's no need to, since there's no data to plot.

I had an initial implementation year ago was filling with nulls between the begining of the query and the first data point. But it was a much more complicated code for no obvious benefit (If I remember the aggregator basically  don't know the time range of the query and some bits have to be modified to get this info available to the aggregators).  

May I ask you what usage you intend to make of those null data points between query start time and first data point?

**Comment by francisdb** [comment]
This was in fact an issue that popped up on client side where we render charts. Those charts take their range from the data that comes from the server so if the user selelect 1-10 jan and there is only data for 4-9 jan he would see a chart for only those 5 days. We have multiple charts on a page and this ended up looking like a big mess as they all showed different ranges.
Of course we were able to fix this issue by correctly configuring the range for the chart drawing library according to the selection of the user.

We migrated from influxdb and that database has a fill option that allows filling each group with a datapoint (null/0/...).
https://docs.influxdata.com/influxdb/v1.1/query_language/data_exploration/#group-by-time-intervals-and-fill

This is hard to implement after querying kairosdb as we have to rebuild the grouping with the same rules.

So we would still like our api to return a datapoint (or null) for each group within the full range.

**Comment by lcoulet** [comment]
I understand, this makes perfect sense. It would be nice to have. I would like Brian and Jeff to give some clues. Making a gaps or fill aggregator that acccepts start time and stop time parameters is not difficult, but it would be better to not require to provide the start/stop time to the aggretaor and pick-up the ones from the query. It's partly why the gaps marking aggregator works that way (and also because the plotting libraries I wa susing were working differently).

You said that you migrated from influxdb, may I ask you what were the reasons that motivated your decision to move away from influxdb? And therefore, what are the reasons to move to KairosDB? What are the challenges of this migration? 
We can discuss this offline if you want.

**Comment by francisdb** [comment]
Influx was just crashing on us all the time, this was in the pre-1.0 period where a lot of writes for past data would trigger huge IO / OOM spikes. Even lost a whole weekend on getting it stable again by reducing non-linear past writes.
Also we needed to scale and they made their clustering a commercial feature.
(we still have an influx instance running but this only for statsd/grafana/internal metrics)

The migration went quite painlessly. One of our interns @thomastoye started writing a Scala KairosDB driver which is available at https://github.com/waylayio/kairosdb-scala and maven central. After which I wrote a custom migration tool for our specific needs.

**Comment by brianhks** [comment]
So changing gaps to take a parameter of what to fill with wouldn't be hard at all.  And from my quick look at the code I'm surprised it isn't filling gaps before and after the data in the query.  Each aggregator gets a startTime from the query, I'll have to dig into it further to understand what is going on.

And thanks for joining the Kairos side of the force :)

**Comment by brianhks** [comment]
I was able to fix the code to fill in the gaps from the beginning of the query to the end.  Change was pretty simple to do.  The question now is should I make the change?

Using it in the Kairos UI makes no difference as the graph drops off all the null values at the beginning and end so, visually it doesn't make a difference.

The fix is in the ExhaustiveRangeDataPointAggregator and one could argue that it was broken because it wasn't truly exhaustive.  I assume the ones replying to this post are the ones that really care about the gaps functionality so what say ye?

**Comment by francisdb** [comment]
I created the ticket so I guess I would like it to be exhaustive :-) Not sure in how far this might break things for existing users?

**Comment by lcoulet** [comment]
Agreed with Francis, better for it to be exhaustive.
Did not want to poke with query JSON parser, but it appears that it was an
easy move for Brian :)





Cordialement,
Loic COULET

2017-01-05 8:46 GMT+00:00 Francis De Brabandere <notifications@github.com>:

**Comment by brianhks** [comment]
OK I've merged this into develop with 342602408fa0043e576cae55237d2d49a9f258fb

**Comment by goekaypamuk** [comment]
it looks like on version 1.2.2 this feature is not working properly anymore.

For me it is crucial that i can specify a range and get exacly an array back with that length. I am trying to plot couple of timeseries on top of eachother and if there is a gap at the start and the end, it has to filled up.  The gaps aggregator is not filling the start and the end. 


```{
  "metrics": [
    {
      "tags": {},
      "name": "M02_10M_WTUR1_W_avg_mag_f",
      "aggregators": [
        {
          "name": "sum",
          "sampling": {
            "value": "10",
            "unit": "minutes"
          }
        },
        {
          "name": "gaps",
          "sampling": {
            "value": "10",
            "unit": "minutes"
          }
        }
      ]
    }
  ],
  "plugins": [],
  "cache_time": 0,
  "start_absolute": 1562796000000,
  "end_absolute": 1562882400000
}```
